File: contracts/dre-contract/src/actions/rename.ts
import { ContractState, ContractInput, ContractResult } from '../types'

export default function rename(state: ContractState, input: ContractInput): ContractResult {
    if (!input.id) throw new ContractError('invalid arguments')

    const caller = SmartWeave.transaction.owner

    const files = state.hierarchy.files[caller]
    
    if (!files) throw new ContractError('drive does not exist')

    const file = files[input.id]

    file.name = input.newName

    return { state }
}

File: contracts/client-contract/src/actions/rename.rs
use crate::types::{ ContractState, Rename };
use warp_contracts::handler_result::WriteResult;

pub fn rename(mut state: ContractState, input: Rename) -> WriteResult<ContractState, String> {
    if !state.is_owner() {
        return WriteResult::ContractError("invalid owner".to_string())
    }

    let file = state.hierarchy.get_file_mut(&input.id);

    if file.is_none() {
        return WriteResult::ContractError("file to rename does not exist".to_string())
    }

    file.unwrap().name = input.new_name;
    WriteResult::Success(state)
}